ibis/pandas/core.py

```diff
@@ -554,3 +610,113 @@ def canonicalize_context(
+        means end1 is earlier than begin2 or end2 is earlier than begin1
+    """
+
+    SUBSET = 0
+    SUPERSET = 1
+    OVERLAP = 2
+    NONOVERLAP = 3
+
+
+def compare_timecontext(cur_context: TimeContext, old_context: TimeContext):
+    """Compare two timecontext and return the relationship between two time
+    context (SUBSET, SUPERSET, OVERLAP, NONOVERLAP).
+
+    Parameters
+    ----------
+    cur_context: TimeContext
+    old_context: TimeContext
+
+    Returns
+    -------
+    result : Enum[TimeContextRelation]
+    """
+    begin, end = cur_context
+    old_begin, old_end = old_context
+    if old_begin <= begin and old_end >= end:
+        return TimeContextRelation.SUBSET
+    elif old_begin >= begin and old_end <= end:
+        return TimeContextRelation.SUPERSET
+    elif old_end < begin or end < old_begin:
+        return TimeContextRelation.NONOVERLAP
+    else:
+        return TimeContextRelation.OVERLAP
+
+
+def scope_item(op, result, timecontext):
+    """make a scope item to be set in scope
+
+    Scope is a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+
+    Parameters
+    ----------
+    op: ibis.expr.operations.Node, key in scope.
+    result : scalar, pd.Series, pd.DataFrame, concrete data.
+    timecontext: Optional[TimeContext], time context associate with the
+    result.
+
+    Returns
+    -------
+    scope_item : Dict, a key value pair that could merge into scope later
+    """
+    return {op: {'value': result, 'timecontext': timecontext}}
+
+
+def get_scope(scope, op, timecontext=None):
+    """ Given a op and timecontext, get result from scope
+    Parameters
+    ----------
+    scope: collections.Mapping
+    a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+    op: ibis.expr.operations.Node, key in scope.
+    timecontext: Optional[TimeContext]
+
+    Returns
+    -------
+    result : scalar, pd.Series, pd.DataFrame
+    """
+    if op not in scope:
+        return None
+    # for ops without timecontext
+    if timecontext is None:
+        return scope[op]['value']
+    else:
+        # For op with timecontext, we only use scope to cache leaf nodes for
+        # now. This is because some ops cannot use cached result with a
+        # different timecontext to get the correct result. For example,
+        # a groupby followed by count, if we use a larger or smaller dataset
+        # from cache, we will probably get an error in result. Such ops with
+        # global aggregation, ops whose result is depending on other rows
+        # in result Dataframe, cannot use cached result with different time
+        # context to optimize calculation. For simplicity, we skip these ops
+        # for now.
+        if isinstance(op, ops.TableColumn) or isinstance(op, ops.TableNode):
```

Comment:
Can you add a xfail test for projection?

Reply:
OK will do

Reply:
Per discussion offline. I think we should just enable this for all nodes instead of leaf nodes.